src/api/app/models/bs_request.rb

```diff
@@ -1033,4 +1052,180 @@ def expand_targets
+    inner_or = []
+    user = User.find_by_login!(user)
+
+    # user's own submitted requests
+    if roles.empty? || roles.include?('creator')
+      inner_or << "bs_requests.creator = #{quote(user.login)}"
+    end
+    # find requests where user is maintainer in target project
+    requests, inner_or = extend_query_for_maintainer(user, requests, roles, inner_or)
+    if roles.empty? || roles.include?('reviewer')
+      requests = requests.includes(:reviews).references(:reviews)
+
+      # requests where the user is reviewer or own requests that are in review by someone else
+      or_in_and = %W(reviews.by_user=#{quote(user.login)})
+
+      # include all groups of user
+      usergroups = user.groups.map { |group| "'#{group.title}'" }
+      or_in_and << "reviews.by_group in (#{usergroups.join(',')})" unless usergroups.blank?
+
+      requests, inner_or = extend_query_for_involved_reviews(user, or_in_and, requests, review_states, inner_or)
+    end
+    if inner_or.empty?
+      requests.none
+    else
+      requests.where(inner_or.join(' or '))
+    end
+  end
+
+  def self.extend_query_for_project(requests, roles, states, review_states, package, subprojects, project)
+    inner_or = []
+    requests, inner_or = extend_relation('source', requests, roles, package, subprojects, project, inner_or)
+    requests, inner_or = extend_relation('target', requests, roles, package, subprojects, project, inner_or)
+
+    if (roles.empty? || roles.include?('reviewer')) &&
+       (states.empty? || states.include?('review'))
+      requests = requests.references(:reviews)
+      review_states.each do |review_state|
+        requests = requests.includes(:reviews)
+        if package.blank?
+          inner_or << "(reviews.state=#{quote(review_state)} and reviews.by_project=#{quote(project)})"
+        else
+          inner_or << "(reviews.state=#{quote(review_state)} and reviews.by_project=#{quote(project)} and reviews.by_package=#{quote(package)})"
+        end
+      end
+    end
+    if inner_or.empty?
+      requests.none
+    else
+      requests.where(inner_or.join(' or '))
+    end
+  end
+
+  def self.extend_relation(source_or_target, requests, roles, package, subprojects, project, inner_or)
+    if roles.empty? || roles.include?(source_or_target)
+      requests = requests.references(:bs_request_actions)
+      if package.blank?
+        if subprojects.blank?
+          inner_or << "bs_request_actions.#{source_or_target}_project=#{quote(project)}"
+        else
+          inner_or << "(bs_request_actions.#{source_or_target}_project like #{quote(project + ':%')})"
+        end
+      else
+        inner_or << "(bs_request_actions.#{source_or_target}_project=#{quote(project)} and " +
+          "bs_request_actions.#{source_or_target}_package=#{quote(package)})"
+      end
+    end
+    [requests, inner_or]
+  end
+
+  def self.extend_query_for_maintainer(obj, requests, roles, inner_or)
+    if roles.empty? || roles.include?('maintainer')
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.